pytorch_lightning/trainer/training_loop.py

```diff
@@ -897,14 +894,20 @@ def should_accumulate(self):
 
     def _should_check_val_fx(self, batch_idx: int, is_last_batch: bool, on_epoch: bool = False) -> bool:
         """ Decide if we should run validation. """
-
         if not self.trainer.enable_validation:
             return False
 
-        # check if this epoch is eligible to run validation
-        if (self.trainer.current_epoch + 1) % self.trainer.check_val_every_n_epoch != 0:
+        is_val_check_epoch = (self.trainer.current_epoch + 1) % self.trainer.check_val_every_n_epoch == 0
+        if not is_val_check_epoch:
             return False
 
+        is_infinite_dataset = self.trainer.val_check_batch == float('inf')
+        if is_last_batch and is_infinite_dataset:
+            return True
+
+        if self.trainer.should_stop:
+            return True
```

Comment:
This is open to discussion, as it previously ran under `if on_epoch`

Reply:
Let's add it to the changelog as it is not strictly a refactor?

Reply:
Actually, going to keep the `on_epoch` guard and address this separately.

```suggestion
        if on_epoch and self.trainer.should_stop:
            return True
```